### PR TITLE
Move to Kotlin

### PR DESCRIPTION
**Pull request by erwinw**
## Move to Kotlin

Move the existing Java implementation to Kotlin

Breaking changes: any breaking changes EXCEPT one: fixing the package for `JacksonParserFactory` (it was mistakenly put as `.guice` instead of `.jackson`).

### CODE REVIEW

File: .github/workflows/build-n-publish.yml [hunk 1]
```diff
@@ -33,10 +33,10 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-gradle-
 
-      - name: Set up JDK 8
+      - name: Set up JDK 17
```
Comment (andersfischernielsen) [inline .github/workflows/build-n-publish.yml:36]: Woah

File: .github/workflows/build-n-publish.yml [hunk 2]
```diff
@@ -54,15 +54,16 @@ jobs:
           ssh-private-key: ${{ secrets.DEPLOY_KEY }}
 
       - name: Semantic Release
-        uses: cycjimmy/semantic-release-action@v2.5.1
+        # v2.7.0 is giving issues
+        uses: cycjimmy/semantic-release-action@v2.6.0
```
Comment (andersfischernielsen) [inline .github/workflows/build-n-publish.yml]: What issues? I'm using it without issue in `jwt-kt` 🤔

**Review by andersfischernielsen — APPROVED** [review]
Initial quick look looks good to me - this is quite a big one to review, so I'm trusting you semi-blindly here @erwinw ❤️‍🔥

File: build.gradle [hunk 3]
```diff
@@ -1,25 +1,57 @@
...
-    repositories {
-        mavenLocal()
-        mavenCentral()
+    tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile) {
+        kotlinOptions {
+            jvmTarget = "11"
+            javaParameters = true
+            languageVersion = "1.6"
+            // See https://blog.jetbrains.com/kotlin/2021/02/the-jvm-backend-is-in-beta-let-s-make-it-stable-together/
+            useIR = true
```
Comment (andersfischernielsen) [inline build.gradle:49]: I'm pretty sure that IR is enabled by default now - I got errors in a recent Gradle DSL version and gave it a look - I can't find the JetBrains docs for the life of me right now though 🤦🏼

File: prop-archaius/src/main/kotlin/io/pleo/prop/archaius/ParsingProperty.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package io.pleo.prop.archaius
+
+import com.netflix.config.PropertyWrapper
+import org.slf4j.LoggerFactory
+import java.util.function.Function
+
+private val logger = LoggerFactory.getLogger(ParsingProperty::class.java)
+
+class ParsingProperty<T>(
+    propName: String?,
```
Comment (andersfischernielsen) [inline prop-archaius/src/main/kotlin/io/pleo/prop/archaius/ParsingProperty.kt:10]: Wouldn't we expect `propName` to always be present here? 🤔

File: prop-core/src/main/kotlin/io/pleo/prop/core/internal/LocalProp.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    override fun invoke() = supplier.get()
+    override fun get() = invoke()
+
+    override val name get() = toString()
+    override val changedTimestamp get() = created
+
+    // Static properties can't change. No callbacks.
+    override fun addCallback(callback: Runnable) {}
+    override fun removeAllCallbacks() {}
```
Comment (andersfischernielsen) [inline prop-core/src/main/kotlin/io/pleo/prop/core/internal/LocalProp.kt:22]: Can we add this comment so it's visible in engineers' editor? Unsure if this is JavaDoc or how IntelliJ pulls out that information - WDYT? Could limit engineers being unpleasantly surprised

File: build.gradle (hunk 3 above)
Comment (erwinw) [reply build.gradle]: OK, removed; thanks!

File: prop-archaius/src/main/kotlin/io/pleo/prop/archaius/ParsingProperty.kt (hunk 4 above)
Comment (erwinw) [reply prop-archaius/src/main/kotlin/io/pleo/prop/archaius/ParsingProperty.kt]: True, fair enough. I've tried to make things that can't be null, not nullable, but some might have slipped through the cracks.

File: .github/workflows/build-n-publish.yml (hunk 2 above)
Comment (erwinw) [reply .github/workflows/build-n-publish.yml]: As you can see in [this run](https://github.com/pleo-io/prop/runs/4868819761?check_suite_focus=true), it fails without any useful info.

**Comment by erwinw** [comment]
> Initial quick look looks good to me - this is quite a big one to review, so I'm trusting you semi-blindly here @erwinw ❤️‍🔥

Trust is good, testing is better -- I've tested this locally by using it with JVM Commons and building & testing it.

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+    private fun parameterToProp(parameter: Parameter, key: Key<*>): Prop<*> {
+        val propertyName: String = getNamedAnnotationValue(parameter.annotations.toList(), key)
+
+        val parameterizedType = parameter.parameterizedType as ParameterizedType
+        val type = parameterizedType.actualTypeArguments.first()
+        val parser = parserFactory.createParserForType(type)
+        try {
+            val annotation = parameter.getAnnotation(Default::class.java)
+            println("ANNOTATION: $annotation")
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:101]: this guy should be removed I think

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                            it.typeLiteral.type is ParameterizedType
+                    } as Key<Prop<*>>?
+                    ?: return@forEach
+
+                val value =
+                    try {
+                        val parameter = parameters[dependency.parameterIndex]
+                        PropResult(parameterToProp(parameter, key))
+                    } catch (ex: RuntimeException) {
+                        PropResult(ex)
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:86]: interesting, how does it work if PropResult will be an exception?

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        val type = parameterizedType.actualTypeArguments.first()
+        val parser = parserFactory.createParserForType(type)
+        try {
+            val annotation = parameter.getAnnotation(Default::class.java)
+            println("ANNOTATION: $annotation")
+            val defaultValue = ofNullable<Default>(annotation)
+                .map(Default::value)
+                .map(parser::apply)
+                .orElse(null)
+            println("DEF: $defaultValue")
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:106]: yeah, and this one

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+            return propFactory.createProp(propertyName, parser, defaultValue)
+        } catch (ex: RuntimeException) {
+            throw FailedToCreatePropException(propertyName, ex)
+        }
+    }
+
+    private fun getNamedAnnotationValue(annotations: List<Annotation>, key: Key<*>): String =
+        annotations.mapNotNull(::annotationValueIfNamed).lastOrNull().also {
+            println("?? $it")
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:116]: one more print to be either removed or logger used

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        }
+            ?: throw RequiredNamedAnnotationException(key)
+
+    private fun annotationValueIfNamed(annotation: Annotation): String? =
+        when (annotation) {
+            is Named -> annotation.value
+            is com.google.inject.name.Named -> annotation.value
+            else -> null
+        }?.ifEmpty { null }.also {
+            println(">> $it")
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:126]: +1

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropResult.kt [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+    constructor(prop: Prop<*>?) {
+        this.prop = prop
+    }
+
+    constructor(error: Throwable?) {
+        this.error = error
+    }
+
+    fun isError() = error != null
```
Comment (andriygg) [inline prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropResult.kt:20]: OK, now I see how it works! 👍

**Review by andriygg — APPROVED** [review]
Great job, E!

File: build.gradle [hunk 12]
```diff
@@ -1,25 +1,55 @@
...
+    ext {
+        kotlinVersion = "1.6.10"
+        kotlinCoroutinesVersion = "1.6.0"
+        kotlinGradlePluginVersion = "1.6.10"
+
+        kotlinterGradleVersion = "3.8.0"
+    }
+
+    repositories {
+        mavenLocal()
```
[the remark this answers is not in the document]
Comment (erwinw) [reply build.gradle:12]: Eh no; Prop is a public repo and does not rely on JVM Commons!

File: prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt (hunk 7 above)
Comment (erwinw) [reply prop-guice/src/main/kotlin/io/pleo/prop/guice/internal/PropMappingVisitor.kt:86]: In a follow-up PR I'm replacing this with the more idiomatic Kotlin [Result](https://thekotlindev.com/2021/kotlin-result-pattern/).

**Pull request merged**